File: cpp/ArrayDeque.cpp
/*
 * ArrayDeque.cpp
 *
 *  Created on: 2011-11-23
 *      Author: morin
 */
#include "utils.h"
#include "ArrayDeque.h"

namespace ods {

template<class T>
ArrayDeque<T>::ArrayDeque() {
	n = 0;
	length = 1;
	a = new T[length];
	j = 0;
}

template<class T>
ArrayDeque<T>::~ArrayDeque() {
	delete a;
}

template<class T>
void ArrayDeque<T>::resize() {
	length = max(1, 2 * n);
	T *b = new T[length];
	for (int k = 0; k < n; k++)
		b[k] = a[(j+k)%length];
	delete a;
	a = b;
}

template<class T>
int ArrayDeque<T>::size()
{
	return n;
}

template<class T>
void ArrayDeque<T>::add(int i, T x) {
	if (n + 1 > length)	resize();
	if (i < n/2) { // shift a[0],..,a[i-1] left one position
		j = (j == 0) ? length - 1 : j - 1;
		for (int k = 0; k <= i-1; k++)
			a[(j+k)%length] = a[(j+k+1)%length];
	} else { // shift a[i],..,a[n-1] right one position
		for (int k = n; k > i; k--)
			a[(j+k)%length] = a[(j+k-1)%length];
	}
	a[(j+i)%length] = x;
	n++;
}


template<class T>
T ArrayDeque<T>::remove(int i)
{
    T x = a[(j+i)%length];
    if (i < n/2) {  // shift a[0],..,[i-1] right one position
    	for (int k = i; k > 0; k--)
			a[(j+k)%length] = a[(j+k-1)%length];
		j = (j + 1) % length;
    } else {        // shift a[i+1],..,a[n-1] left one position
		for (int k = i; k < n-1; k++)
			a[(j+k)%length] = a[(j+k+1)%length];
    }
    n--;
    if (3*n < length) resize();
    return x;
}

template ArrayDeque<int>::ArrayDeque();
template ArrayDeque<int>::~ArrayDeque();
template void ArrayDeque<int>::add(int,int);
template int ArrayDeque<int>::remove(int);
template int ArrayDeque<int>::size();

} /* namespace ods */

File: python/arraydeque.py
'''
Created on 2012-04-02

@author: morin
'''

        
class ArrayDeque(object):
    def __init__(self):
        self.a = [None]
        self.j = 0
        self.n = 0

    def get(self, i):
        if (i < 0 or i >= self.n): raise IndexError()
        return self.a[(i+self.j)%len(self.a)]

    def set(self, i, x): #@ReservedAssignment
        if (i < 0 or i >= self.n): raise IndexError()
        y = self.a[(i+self.j)%len(self.a)]
        self.a[(i+self.j)%len(self.a)] = x
        return y

    def add(self, i, x): 
        if (i < 0 or i > self.n): raise IndexError()
        if (self.n == len(self.a)): self._resize()
        if (i < self.n/2):
            self.j = (self.j-1) % len(self.a)
            for k in xrange(i):
                self.a[(self.j+k)%len(self.a)] = self.a[(self.j+k+1)%len(self.a)]
        else:
            for k in xrange(self.n, i, -1):
                self.a[(self.j+k)%len(self.a)] = self.a[(self.j+k-1)%len(self.a)]
        self.a[(self.j+i)%len(self.a)] = x
        self.n += 1

    def remove(self, i): 
        if (i < 0 or i >= self.n): raise IndexError()
        x = self.a[(self.j+i)%len(self.a)]
        if (i < self.n / 2):
            for k in xrange(i, 0, -1):
                self.a[(self.j+k)%len(self.a)] = self.a[(self.j+k-1)%len(self.a)]
            self.j = (self.j+1) % len(self.a)
        else:
            for k in xrange(i, self.n-1): 
                self.a[(self.j+k)%len(self.a)] = self.a[(self.j+k+1)%len(self.a)]
        self.n -= 1
        if (len(self.a) >= 3*self.n): self._resize();
        return x;
  
    def _resize(self):
        b = [None] * max(1, 2*self.n)
        for k in xrange(self.n):
            b[k] = self.a[(self.j+k)%len(self.a)]
        self.a = b
        self.j = 0
        
    def size(self):
        return self.n
    
    def __len__(self):
        return self.size()
    
    def __str__(self):
        s = "["
        for i in xrange(self.n):
            s += self.get(i)
            if i < self.n-1: s += ","
        s += "]"
        
